### PR TITLE
Forbid orphan components

### PR DESCRIPTION
**Pull request by pmvald**
This PR adds an option to Angular compiler options which allow to produce runtime error if orphan components found. Good for identifying such cases. 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts [hunk 1]
```diff
@@ -262,6 +262,12 @@ export interface DiagnosticOptions {
      */
     checks?: {[Name in ExtendedTemplateDiagnosticName]?: DiagnosticCategoryLabel};
   };
+
+  /**
+   * Enables the runtime check to guard against rendering a component without first loading its
+   * NgModule.
+   */
+  forbidOrphanComponents?: boolean;
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:270]: Wondering if we should add this to `MiscOptions` instead. We are not issuing any compiler build-time diagnostic

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:270]: Maybe also we should clarify that this option will only prevent orphan use of components _from this compilation_. i.e. the use of a component from a library that is not compiled with that option will _not_ have such runtime check, right?

File: packages/compiler/src/render3/r3_class_debug_info_compiler.ts [hunk 2]
```diff
@@ -41,24 +41,39 @@ export interface R3ClassDebugInfo {
...
+    className: debugInfo.className,
+  };
 
   // Include file path and line number only if the file relative path is calculated successfully.
   if (debugInfo.filePath) {
     debugInfoObject.filePath = debugInfo.filePath;
     debugInfoObject.lineNumber = debugInfo.lineNumber;
   }
 
+  // Include forbidOrphanRendering only if it's set to true (to reduce the writing work!)
```
Comment (devversion) [inline packages/compiler/src/render3/r3_class_debug_info_compiler.ts:72]:
```suggestion
  // Include forbidOrphanRendering only if it's set to true (to reduce generated code).
```

File: packages/router/src/state_manager.ts [hunk 3]
```diff
@@ -106,8 +106,8 @@ export class StateManager {
     };
   }
 
-  nonRouterCurrentEntryChange(
-      listener: (url: string, state: RestoredState|null|undefined) => void): SubscriptionLike {
+  nonRouterCurrentEntryChange(listener: (url: string, state: RestoredState|null|undefined) => void):
+      SubscriptionLike {
```
Comment (devversion) [inline packages/router/src/state_manager.ts:110]: Nit: Unrelated change?

File: packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts [hunk 4]
```diff
@@ -481,7 +482,9 @@ export class ComponentDecoratorHandler implements
                 node, this.reflector, this.isCore, this.annotateForClosureCompiler,
                 dec => transformDecoratorResources(dec, component, styles, template)) :
             null,
-        classDebugInfo: extractClassDebugInfo(node, this.reflector, this.rootDirs),
+        classDebugInfo: extractClassDebugInfo(
+            node, this.reflector, this.rootDirs,
+            !metadata.isStandalone && this.forbidOrphanRendering),
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:487]: for more clarity / or we use a keyed options object (same in the test)

```suggestion
           /* forbidOrphanRenderering */ !metadata.isStandalone && this.forbidOrphanRendering),
```

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:487]: Do we just want to pass the option and let runtime deal with the standalone case? (that would seem more consisten to me)

File: packages/core/src/render3/component_ref.ts [hunk 5]
```diff
@@ -163,6 +164,17 @@ export class ComponentFactory<T> extends AbstractComponentFactory<T> {
       injector: Injector, projectableNodes?: any[][]|undefined, rootSelectorOrNode?: any,
       environmentInjector?: NgModuleRef<any>|EnvironmentInjector|
       undefined): AbstractComponentRef<T> {
+    // Check if the component is orphan
+    if (ngDevMode && this.componentDef.debugInfo?.forbidOrphanRendering) {
+      if (depsTracker.isOrphanComponent(this.componentType)) {
```
Comment (devversion) [inline packages/core/src/render3/component_ref.ts:169]: There is an interesting scenario we should chat about (related to our ngDevMode/jitMode conversations). If `ngJitMode` is `false` but `ngDevMode` is true, this will currently always yield an error IIRC because `setNgModuleScope` would never be invoked?

File: packages/core/src/render3/component_ref.ts [hunk 6]
```diff
@@ -163,6 +164,17 @@ export class ComponentFactory<T> extends AbstractComponentFactory<T> {
...
+        throw new RuntimeError(
+            RuntimeErrorCode.RUNTIME_DEPS_ORPHAN_COMPONENT,
+            `Orphan component found! Trying to render the component ${
+                debugStringifyTypeForError(
+                    this.componentType)} without first loading the NgModule that declares it. It is recommended to make this component standalone in order to avoid such confusing cases. If this is not possible now, please import the component's NgModule in the NgModule or the standalone component in which you are trying to render this component. Also make sure the way the app is bundled and served always includes the component's NgModule before the component.`);
```
Comment (devversion) [inline packages/core/src/render3/component_ref.ts:174]:
```suggestion
                    this.componentType)} without first loading the NgModule that declares it. It is recommended to make this component standalone in order to avoid this error. If this is not possible now, import the component's NgModule in the appropriate NgModule, or the standalone component in which you are trying to render this component. If this is a lazy import, load the NgModule lazily as well and use its module injector.`);
```

I think it's important to call out the solution for lazy NgModule cases (which are not super uncommon). I'd drop the order of bundling. Feels a little too specific and most Angular users don't control bundling

**Review by devversion — COMMENTED** [review]
Some initial comments.

File: packages/core/src/render3/component_ref.ts (hunk 5 above)
Comment (devversion) [reply packages/core/src/render3/component_ref.ts:170]: Sounds good. My first thought is that `ngDebugMode` should _not_ exist and `ngDevMode` should be sufficient. We are not adding any super verbose logging/debugging as far as I know

**Review by devversion — APPROVED** [review]

**Review by crisbeto — APPROVED** [review]

**Comment by pmvald** [comment]
Caretaker note: the ci/test failed due to `//packages/core/test/bundling/defer:symbol_test` which has been flaky these days. But it is fixed this morning and once rebased, this PR should pass that as well.

**Comment by atscott** [comment]
This PR was merged into the repository by commit 59ba2a6e9f7218c78cd8b5f4c13ffb46dd0a4900.
